crates/sui-core/src/execution_scheduler/mod.rs

```diff
@@ -0,0 +1,1115 @@
+        tx_ready_certificates: UnboundedSender<PendingCertificate>,
+        metrics: Arc<AuthorityMetrics>,
+    ) -> Self {
+        Self {
+            object_cache_read,
+            transaction_cache_read,
+            pending_certificates: Arc::new(RwLock::new(HashSet::new())),
+            overload_tracker: Arc::new(OverloadTracker::new()),
+            tx_ready_certificates,
+            metrics,
+        }
+    }
+
+    pub(crate) fn enqueue_with_expected_effects_digest(
+        &self,
+        certs: Vec<(VerifiedExecutableTransaction, TransactionEffectsDigest)>,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) {
+        let certs = certs
+            .into_iter()
+            .map(|(cert, fx)| (cert, Some(fx)))
+            .collect();
+        self.enqueue_impl(certs, epoch_store)
+    }
+
+    pub(crate) fn enqueue_certificates(
+        &self,
+        certs: Vec<VerifiedCertificate>,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) {
+        let executable_txns = certs
+            .into_iter()
+            .map(VerifiedExecutableTransaction::new_from_certificate)
+            .collect();
+        self.enqueue(executable_txns, epoch_store)
+    }
+
+    pub(crate) fn enqueue(
+        &self,
+        certs: Vec<VerifiedExecutableTransaction>,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) {
+        let certs = certs.into_iter().map(|cert| (cert, None)).collect();
+        self.enqueue_impl(certs, epoch_store)
+    }
+
+    fn enqueue_impl(
+        &self,
+        certs: Vec<(
+            VerifiedExecutableTransaction,
+            Option<TransactionEffectsDigest>,
+        )>,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) {
+        let certs = certs.into_iter().filter_map(|cert| {
+            if cert.0.epoch() == epoch_store.epoch() {
+                Some(cert)
+            } else {
+                warn!(
+                    "Ignoring enqueued certificate from wrong epoch. Expected={} Certificate={:?}",
+                    epoch_store.epoch(),
+                    cert.0.epoch(),
+                );
+                None
+            }
+        });
+
+        for cert in certs {
+            let scheduler = self.clone();
+            let epoch_store = epoch_store.clone();
+            spawn_monitored_task!(
+                epoch_store.within_alive_epoch(scheduler.schedule_transaction(
+                    cert.0,
+                    cert.1,
+                    &epoch_store,
+                ))
+            );
+        }
+
+        self.metrics
+            .transaction_manager_num_pending_certificates
+            .set(self.pending_certificates.read().len() as i64);
+    }
+
+    async fn schedule_transaction(
+        self,
+        cert: VerifiedExecutableTransaction,
+        expected_effects_digest: Option<TransactionEffectsDigest>,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) {
+        if !self.pending_certificates.write().insert(*cert.digest()) {
```

Comment:
its not clear to me that pending_certificates is still necessary. it seems to occasionally prevent a double execution, but that should be a rare case. it may be better to remove it and just rely on the notify_read_effects

Reply:
It's also used for overload control, although that part could in theory be replaced with an atomic counter, but since we need something there, this seems ok-ish.